sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1197 @@
+
+A graph in the executable state can have each nodes inputs & outputs updated
+using the `command_graph::update()` method. This takes a graph in the
+modifiable state and updates the executable graph to use the node input &
+outputs of the modifiable graph, a technique called _Whole Graph Update_. The
+modifiable graph must have the same topology as the graph originally used to
+create the executable graphs, with the nodes added in the same order.
+
+==== Graph Member Functions
+
+Table 5. Constructor of the `command_graph` class.
+[cols="2a,a"]
+|===
+|Constructor|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+command_graph(const property_list& propList = {});
+----
+|Creates a SYCL `command_graph` object in the modifiable state.
+Zero or more properties can be provided to the constructed SYCL `command_graph`
+via an instance of `property_list`.
+
+Preconditions:
+
+* This constructor is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `propList` - Optional parameter for passing properties. No `command_graph`
+  constructor properties are defined by this extension.
+
+|===
+
+Table 6. Member functions of the `command_graph` class.
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+node add(const property_list& propList = {});
+----
+|This creates an empty node which contains no command. Its intended use is
+either a connection point inside a graph between groups of nodes, and can
+significantly reduce the number of edges ( O(n) vs. O(n^2) ). Another use-case
+is building the structure of a graph first and adding tasks later.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `propList` - Zero or more properties can be provided to the constructed node
+  via an instance of `property_list`.
+
+Returns: The empty node which has been added to the graph.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+template<typename T>
+node add(T cgf, const property_list& propList = {});
+----
+|This function adds a command group function object to a graph. The function
+object statically contains a group of commands, of which a single command is
+executed at runtime. A function object can be a host task which is scheduled by
+the SYCL runtime, or a SYCL function for invoking kernels with all restrictions
+that apply as described in the core specification. The requisites of `cgf` will
+be used to identify any dependent nodes in the graph to form edges with.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `cgf` - Command group function object to be added as a node.
```

Comment:
I was looking at all the member functions for `sycl::handler` to see if they all make sense for graph nodes.  Will specialization constants be a problem?  What happens if one node in the graph sets the value of specialization constant X to one value and another node in the same graph sets the value of that same specialization constant X to something else?  Can we handle a case like this?

Do you foresee a problem using any of the other member functions for a graph node?  If there are functions that don't make sense for graphs, we could consider changing the definition of `command_graph::add`, so that the `cgf` function had a different signature, like:

```
commandgraph.add([&](sycl::ext::oneapi::experimental::graph_handler &cgh) {
  // ...
});
```

This would allow us to define `graph_handler` so it only exposes features that make sense for graph nodes.
```

Reply:
I don't think that spec constants should present much of a problem. On the specification side I can't see anything that affects this kind of use case. On the implementation side we are still dealing with individual kernels which can have their spec constants set in the normal way through SPIR-V, or in the case of emulated spec constants their kernel parameters set as normal. In the emulated case I believe the buffer used to store spec constant values is unique for each submission to prevent overwriting values, so this should not present problems even when batching kernels together for submission to a device.

I believe most if not all handler functions should be able to be supported by graphs as well so I'm not sure a unique handler would be necessary in this case. It would also present serious problems for record & replay particularly in library calls where the handler used for submission can't be changed. Keeping the existing mechanisms as much as possible also helps users porting code to the explicit API since it reduces the changes required and potential for errors.